src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -199,33 +198,90 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    Task socketTask = socket.ConnectAsync(address, port);
+                    socketTasks.Add(socketTask, socket);
                 }
                 catch (Exception exc)
                 {
                     socket.Dispose();
-                    lastException = ExceptionDispatchInfo.Capture(exc);
+                    lastSocketConnectEx = ExceptionDispatchInfo.Capture(exc);
+                }
+            }
+
+            if (socketTasks.Count == 0)
+            {
+                if (lastSocketConnectEx != null)
+                {
+                    lastSocketConnectEx.Throw();
+                }
+                else
+                {
+                    // Should never get here. Either there will have been no addresses and we'll have thrown
+                    // at the beginning, or one of the addresses will have worked and been added to the list, or
+                    // at least one of the addresses will have failed, raising an exception.
+                    throw new ArgumentException();
+                }
+            }
+
+            // Wait on all connection tasks, and return socket associated with first completed task.
+            Socket resultSocket = null;
+            ExceptionDispatchInfo connectWaitEx = null;
+            try
+            {
+                while (socketTasks.Count > 0)
+                {
+                    Task completedTask = await Task.WhenAny(socketTasks.Keys).ConfigureAwait(false);
+                    Socket completedSocket = socketTasks[completedTask];
+
+                    if (completedSocket.Connected)
+                    {
+                        resultSocket = completedSocket;
+                        // Remove successful socket from dictionary so that it doesn't get Disposed in finally block.
+                        socketTasks.Remove(completedTask);
+                        break;
+                    }
+                    else
+                    {
+                        // Remove failed socket and its associated task and continue.
+                        completedSocket.Dispose();
+                        socketTasks.Remove(completedTask);
+                    }
+                }
+            }
+            catch (Exception exc)
+            {
+                connectWaitEx = ExceptionDispatchInfo.Capture(exc);
+            }
+            finally
+            {
+                // Dispose any sockets that were still pending at time of success/error
+                foreach (Socket socket in socketTasks.Values)
```

Comment:
Can this clean up be done asynchronously, so that you can return the good connection to caller as soon as it is available?